File: server/src/models/Transaction.ts
import { Sequelize, DataTypes, Model, ModelStatic, Optional } from 'sequelize';
import { IUserModel } from './User';
import { IChitNoteModel } from './ChitNote';

export interface ITransactionAttributes {
  id: number;
  date: Date;
  receiptNo: string;
  amount: number;
  totalAmount: number;
  total: number | null;
  userId: number;
  chitNoteId: number;
  createdAt?: Date;
  updatedAt?: Date;
  user?: IUserModel;
  chitNote?: IChitNoteModel;
}

export interface ITransactionCreationAttributes extends Optional<ITransactionAttributes, 'id' | 'total' | 'createdAt' | 'updatedAt'> {}

export interface ITransactionModel extends Model<ITransactionAttributes, ITransactionCreationAttributes>, ITransactionAttributes {
  // Instance methods can be added here
}

type TransactionModelStatic = ModelStatic<ITransactionModel> & {
  associate?: (models: {
    User: typeof IUserModel;
    ChitNote: typeof IChitNoteModel;
  }) => void;
};

export const transactionModel = (sequelize: Sequelize): TransactionModelStatic => {
  const Transaction = sequelize.define<ITransactionModel>('Transaction', {
    id: {
      type: DataTypes.INTEGER,
      primaryKey: true,
      autoIncrement: true,
      allowNull: false
    },
    date: {
      type: DataTypes.DATE,
      allowNull: false,
      field: 'date',
      defaultValue: DataTypes.NOW
    },
    receiptNo: {
      type: DataTypes.STRING(100),
      allowNull: false,
      field: 'receipt_no',
      unique: true,
      validate: {
        notEmpty: true
      }
    },
    amount: {
      type: DataTypes.DECIMAL(10, 2),
      allowNull: false,
      field: 'amount',
      validate: {
        min: 0
      }
    },
    totalAmount: {
      type: DataTypes.DECIMAL(10, 2),
      allowNull: false,
      field: 'total_amount',
      validate: {
        min: 0
      }
    },
    total: {
      type: DataTypes.DECIMAL(10, 2),
      allowNull: true,
      field: 'total',
      validate: {
        min: 0
      }
    },
    userId: {
      type: DataTypes.INTEGER,
      allowNull: false,
      field: 'user_id',
      references: {
        model: 'users',
        key: 'id'
      }
    },
    chitNoteId: {
      type: DataTypes.INTEGER,
      allowNull: false,
      field: 'chit_note_id',
      references: {
        model: 'chit_notes',
        key: 'id'
      }
    },
    createdAt: {
      type: DataTypes.DATE,
      allowNull: false,
      field: 'created_at',
      defaultValue: DataTypes.NOW
    },
    updatedAt: {
      type: DataTypes.DATE,
      allowNull: false,
      field: 'updated_at',
      defaultValue: DataTypes.NOW
    }
  }, {
    tableName: 'transactions',
    timestamps: true,
    underscored: true,
    indexes: [
      {
        fields: ['user_id']
      },
      {
        fields: ['chit_note_id']
      },
      {
        unique: true,
        fields: ['receipt_no']
      }
    ]
  }) as unknown as TransactionModelStatic;

  // Define associations
  Transaction.associate = (models) => {
    Transaction.belongsTo(models.User, {
      foreignKey: 'userId',
      as: 'user',
      onDelete: 'CASCADE'
    });
    
    Transaction.belongsTo(models.ChitNote, {
      foreignKey: 'chitNoteId',
      as: 'chitNote',
      onDelete: 'CASCADE'
    });
  };

  return Transaction;
};

export default transactionModel;

File: serverPy/models/transaction.py
from config.database import db
from models.base_model import BaseModel

class Transaction(BaseModel):
    __tablename__ = 'transactions'
    
    id = db.Column(db.Integer, primary_key=True)
    date = db.Column(db.DateTime, nullable=False, default=db.func.current_timestamp())
    receipt_no = db.Column(db.String(100), unique=True, nullable=False)
    amount = db.Column(db.Numeric(10, 2), nullable=False)
    total_amount = db.Column(db.Numeric(10, 2), nullable=False)
    total = db.Column(db.Numeric(10, 2), nullable=True)
    user_id = db.Column(db.Integer, db.ForeignKey('users.id', ondelete='CASCADE'), nullable=False)
    chit_note_id = db.Column(db.Integer, db.ForeignKey('chit_notes.id', ondelete='CASCADE'), nullable=False)
    
    def __init__(self, receipt_no, amount, total_amount, user_id, chit_note_id, total=None, date=None):
        self.receipt_no = receipt_no
        self.amount = amount
        self.total_amount = total_amount
        self.total = total
        self.user_id = user_id
        self.chit_note_id = chit_note_id
        if date:
            self.date = date
    
    def to_dict(self):
        return {
            'id': self.id,
            'date': self.date.isoformat() if self.date else None,
            'receipt_no': self.receipt_no,
            'amount': float(self.amount) if self.amount is not None else None,
            'total_amount': float(self.total_amount) if self.total_amount is not None else None,
            'total': float(self.total) if self.total is not None else None,
            'user_id': self.user_id,
            'chit_note_id': self.chit_note_id,
            'created_at': self.created_at.isoformat() if self.created_at else None,
            'updated_at': self.updated_at.isoformat() if self.updated_at else None
        }
    
    def __repr__(self):
        return f'<Transaction {self.receipt_no}>'
